File: ontos/src/frontend/src/types/costs.ts
export type EntityKind = 'data_domain' | 'data_product' | 'data_contract';

export type CostCenter = 'INFRASTRUCTURE' | 'HR' | 'STORAGE' | 'MAINTENANCE' | 'OTHER';

export interface CostItem {
  id: string;
  entity_id: string;
  entity_type: EntityKind;
  title?: string | null;
  description?: string | null;
  cost_center: CostCenter;
  custom_center_name?: string | null;
  amount_cents: number;
  currency: string; // ISO 4217
  start_month: string; // YYYY-MM-DD
  end_month?: string | null; // YYYY-MM-DD
  created_at: string;
  updated_at: string;
}

export interface CostItemCreate {
  entity_id: string;
  entity_type: EntityKind;
  title?: string | null;
  description?: string | null;
  cost_center: CostCenter;
  custom_center_name?: string | null;
  amount_cents: number;
  currency: string;
  start_month: string; // YYYY-MM-DD
  end_month?: string | null; // YYYY-MM-DD
}

export interface CostItemUpdate extends Partial<Omit<CostItemCreate, 'entity_id' | 'entity_type'>> {}

export interface CostSummary {
  month: string; // YYYY-MM
  currency: string;
  total_cents: number;
  items_count: number;
  by_center: Record<CostCenter | string, number>;
}

export const formatCents = (amountCents: number, currency: string) => {
  try {
    return new Intl.NumberFormat(undefined, { style: 'currency', currency }).format(
      (amountCents || 0) / 100
    );
  } catch {
    return `$${((amountCents || 0) / 100).toFixed(2)}`;
  }
};

File: ontos/src/backend/src/models/costs.py
from __future__ import annotations
from typing import Optional, Dict
from uuid import UUID
from datetime import date, datetime

from pydantic import BaseModel, Field


class CostCenter(str):
    INFRASTRUCTURE = "INFRASTRUCTURE"
    HR = "HR"
    STORAGE = "STORAGE"
    MAINTENANCE = "MAINTENANCE"
    OTHER = "OTHER"


class CostItemBase(BaseModel):
    entity_id: str
    entity_type: str = Field(..., pattern=r"^(data_domain|data_product|data_contract)$")
    title: Optional[str] = Field(None, max_length=255)
    description: Optional[str] = Field(None, max_length=2000)
    cost_center: str = Field(..., pattern=r"^(INFRASTRUCTURE|HR|STORAGE|MAINTENANCE|OTHER)$")
    custom_center_name: Optional[str] = Field(None, max_length=255)
    amount_cents: int = Field(..., ge=0)
    currency: str = Field("USD", min_length=3, max_length=3)
    start_month: date
    end_month: Optional[date] = None


class CostItemCreate(CostItemBase):
    pass


class CostItemUpdate(BaseModel):
    title: Optional[str] = Field(None, max_length=255)
    description: Optional[str] = Field(None, max_length=2000)
    cost_center: Optional[str] = Field(None, pattern=r"^(INFRASTRUCTURE|HR|STORAGE|MAINTENANCE|OTHER)$")
    custom_center_name: Optional[str] = Field(None, max_length=255)
    amount_cents: Optional[int] = Field(None, ge=0)
    currency: Optional[str] = Field(None, min_length=3, max_length=3)
    start_month: Optional[date] = None
    end_month: Optional[date] = None


class CostItem(CostItemBase):
    id: UUID
    created_by: Optional[str] = None
    updated_by: Optional[str] = None
    created_at: datetime
    updated_at: datetime

    model_config = {
        "from_attributes": True
    }


class CostSummary(BaseModel):
    month: str  # YYYY-MM
    currency: str
    total_cents: int
    items_count: int
    by_center: Dict[str, int]


